File: src/utils/grading.py
import pandas as pd


def calculate_grade_info(pct: float, grading_df: pd.DataFrame = None) -> dict:
    """
    Computes Letter Grade, Remarks, and Pass/Fail status based on Percentage.
    Leverages Grading_System threshold mappings if defined, or falls back to standard thresholds.
    """
    pct = round(float(pct), 2)

    # Check if threshold columns exist in Grading_System
    if grading_df is not None and not grading_df.empty:
        min_col = "Min_Percentage" if "Min_Percentage" in grading_df.columns else ("Min Percentage" if "Min Percentage" in grading_df.columns else None)
        max_col = "Max_Percentage" if "Max_Percentage" in grading_df.columns else ("Max Percentage" if "Max Percentage" in grading_df.columns else None)

        if min_col and max_col:
            g_df = grading_df.copy()
            g_df["Min_Pct"] = pd.to_numeric(g_df[min_col].astype(str).str.replace("%", ""), errors="coerce")
            g_df["Max_Pct"] = pd.to_numeric(g_df[max_col].astype(str).str.replace("%", ""), errors="coerce")
            g_df = g_df.dropna(subset=["Min_Pct", "Max_Pct"])

            for _, row in g_df.iterrows():
                if row["Min_Pct"] <= pct <= row["Max_Pct"]:
                    grade = str(row.get("Grade", "")).strip() or "N/A"
                    remarks = str(row.get("Remarks", "")).strip() or "Satisfactory"
                    status = "PASS" if grade not in ["F", "Fail", "FAIL", "U"] and pct >= 40 else "FAIL"
                    return {"grade": grade, "remarks": remarks, "status": status}

    # Standard Fallback Thresholds (PSCC Academic Standard)
    if pct >= 95:
        return {"grade": "A++", "remarks": "Exceptional", "status": "PASS"}
    elif pct >= 90:
        return {"grade": "A+", "remarks": "Outstanding", "status": "PASS"}
    elif pct >= 85:
        return {"grade": "A", "remarks": "Excellent", "status": "PASS"}
    elif pct >= 80:
        return {"grade": "B++", "remarks": "Very Good", "status": "PASS"}
    elif pct >= 75:
        return {"grade": "B+", "remarks": "Good", "status": "PASS"}
    elif pct >= 70:
        return {"grade": "B", "remarks": "Fairly Good", "status": "PASS"}
    elif pct >= 60:
        return {"grade": "C", "remarks": "Above Average", "status": "PASS"}
    elif pct >= 50:
        return {"grade": "D", "remarks": "Average", "status": "PASS"}
    elif pct >= 40:
        return {"grade": "E", "remarks": "Below Average", "status": "PASS"}
    else:
        return {"grade": "U", "remarks": "Fail / Unsatisfactory", "status": "FAIL"}

File: lib/grading.js
const STANDARD_GRADING_SCALE = Object.freeze([
  { min: 95, max: 100, grade: "A++", remarks: "Exceptional" },
  { min: 90, max: 94.999999, grade: "A+", remarks: "Outstanding" },
  { min: 85, max: 89.999999, grade: "A", remarks: "Excellent" },
  { min: 80, max: 84.999999, grade: "B++", remarks: "Very Good" },
  { min: 75, max: 79.999999, grade: "B+", remarks: "Good" },
  { min: 70, max: 74.999999, grade: "B", remarks: "Fairly Good" },
  { min: 60, max: 69.999999, grade: "C", remarks: "Above Average" },
  { min: 50, max: 59.999999, grade: "D", remarks: "Average" },
  { min: 40, max: 49.999999, grade: "E", remarks: "Below Average" },
  { min: 0, max: 39.999999, grade: "U", remarks: "Fail / Unsatisfactory" },
]);

function parseBoundary(value) {
  const raw = String(value ?? "").trim().replace(/%$/, "").trim();
  if (!/^\d+(?:\.\d+)?$/.test(raw)) return null;
  const number = Number(raw);
  return Number.isFinite(number) ? number : null;
}

function configuredScale(gradingSystem) {
  if (!Array.isArray(gradingSystem) || gradingSystem.length === 0) return null;
  const rows = gradingSystem.map((row) => ({
    min: parseBoundary(row.Min_Percentage ?? row["Min Percentage"]),
    max: parseBoundary(row.Max_Percentage ?? row["Max Percentage"]),
    grade: String(row.Grade ?? "").trim(),
    remarks: String(row.Remarks ?? "").trim() || "Satisfactory",
  }));
  if (rows.some((row) => row.min === null || row.max === null || row.min > row.max || !row.grade)) {
    return { valid: false, rows: [], error: "The configured grading scale contains invalid boundaries." };
  }
  rows.sort((left, right) => right.min - left.min || right.max - left.max || left.grade.localeCompare(right.grade));
  for (let index = 1; index < rows.length; index++) {
    if (rows[index].max >= rows[index - 1].min) {
      return { valid: false, rows: [], error: "The configured grading scale contains overlapping boundaries." };
    }
  }
  return { valid: true, rows, error: null };
}

export function getGradingScale(gradingSystem = []) {
  return configuredScale(gradingSystem) || { valid: true, rows: STANDARD_GRADING_SCALE, error: null };
}

export function calculateGradeInfo(pct, gradingSystem = []) {
  const percentage = Math.round(Number(pct) * 100) / 100;
  if (!Number.isFinite(percentage) || percentage < 0 || percentage > 100) {
    return { grade: "N/A", remarks: "-", status: "-", valid: false };
  }
  const scale = getGradingScale(gradingSystem);
  if (!scale.valid) {
    return { grade: "N/A", remarks: scale.error, status: "-", valid: false };
  }
  const match = scale.rows.find((row) => percentage >= row.min && percentage <= row.max);
  if (!match) {
    return {
      grade: "N/A",
      remarks: "The configured grading scale does not cover this percentage.",
      status: "-",
      valid: false,
    };
  }
  const gradeKey = match.grade.toUpperCase();
  const isFail = ["F", "FAIL", "U"].includes(gradeKey) || percentage < 40;
  return {
    grade: match.grade,
    remarks: match.remarks,
    status: isFail ? "FAIL" : "PASS",
    valid: true,
  };
}

export { STANDARD_GRADING_SCALE };
